src/Compilers/CSharp/Portable/Symbols/Source/SourceMethodSymbolWithAttributes.cs

```diff
@@ -808,6 +812,88 @@ private void DecodeModuleInitializerAttribute(DecodeWellKnownAttributeArguments<
                 DeclaringCompilation.AddModuleInitializerMethod(this);
             }
         }
+
+        private void DecodeUnmanagedCallersOnlyAttribute(DecodeWellKnownAttributeArguments<AttributeSyntax, CSharpAttributeData, AttributeLocation> arguments)
+        {
+            Debug.Assert(arguments.AttributeSyntaxOpt is not null);
+            if (!IsStatic || MethodKind is not (MethodKind.Ordinary or MethodKind.LocalFunction))
+            {
+                // `UnmanagedCallersOnly` can only be applied to ordinary static methods or local functions.
+                arguments.Diagnostics.Add(ErrorCode.ERR_UnmanagedCallersOnlyRequiresStatic, arguments.AttributeSyntaxOpt.Location);
+            }
+
+            if (!arguments.Attribute.CommonNamedArguments.IsDefaultOrEmpty)
+            {
+                foreach (var (key, value) in arguments.Attribute.CommonNamedArguments)
+                {
+                    if (key != "CallConvs"
+                        || value.Kind != TypedConstantKind.Array
+                        || value.Values.Any(v => v.Kind != TypedConstantKind.Type))
+                    {
+                        continue;
+                    }
+
+                    foreach (var callConvTypedConstant in value.Values)
+                    {
+                        Debug.Assert(callConvTypedConstant.Kind == TypedConstantKind.Type);
+                        if (!(callConvTypedConstant.ValueInternal is NamedTypeSymbol callConvType)
+                            || !FunctionPointerTypeSymbol.IsCallingConventionModifier(callConvType))
+                        {
+                            // `{0}` is not a valid calling convention type for 'UnmanagedCallersOnly'.
+                            arguments.Diagnostics.Add(ErrorCode.ERR_InvalidUnmanagedCallersOnlyCallConv, arguments.AttributeSyntaxOpt.Location, callConvTypedConstant.ValueInternal ?? "null");
+                        }
+                    }
+                }
+            }
+
+            var returnTypeSyntax = SyntaxNode switch
+            {
+                MethodDeclarationSyntax m => m.ReturnType,
+                LocalFunctionStatementSyntax l => l.ReturnType,
+                _ => null
+            };
+
+            if (returnTypeSyntax is null)
```

Comment:
>returnTypeSyntax is null [](start = 16, length = 24)

Not sure when `returnTypeSyntax` would be `null`, but would it make sense to just skip the `ReturnType` and still check `Parameters` in this case?

Reply:
The `ReturnTypeSyntax` will only be `null` if this wasn't an ordinary method or local function. For these cases, we've already reported an error above that the attribute is not valid, anything further would not be useful imo.

---
In reply to: [476630383](https://github.com/dotnet/roslyn/pull/47008#discussion_r476630383) [](ancestors = 476630383)